src/modules/DetectionSensorModule.cpp

```diff
@@ -0,0 +1,94 @@
+
+    if (moduleConfig.detection_sensor.enabled == false)
+        return disable();
+
+    if (firstTime) {
+        // This is the first time the OSThread library has called this function, so do some setup
+        firstTime = false;
+        if (moduleConfig.detection_sensor.monitor_pin > 0) {
+            pinMode(moduleConfig.detection_sensor.monitor_pin, moduleConfig.detection_sensor.use_pullup ? INPUT_PULLUP : INPUT);
+        }
+        LOG_INFO("Detection Sensor Module: Initializing\n");
+
+        return DELAYED_INTERVAL;
+    }
+
+    if ((millis() - lastSentToMesh) >= getConfiguredOrDefaultMs(moduleConfig.detection_sensor.minimum_broadcast_secs) &&
+        hasStateChanged()) {
+        sendDetectionMessage();
+        return DELAYED_INTERVAL;
+    }
+    // Even if we haven't detected a change from state, broadcast our current state the mesh on the scheduled interval as a sort
+    // of heartbeat. We only do this if the minimum broadcast interval is greater than zero, otherwise we'll only broadcast state
+    // change detections.
+    else if (moduleConfig.detection_sensor.state_broadcast_secs > 0 &&
+             (millis() - lastSentToMesh) >= getConfiguredOrDefaultMs(moduleConfig.detection_sensor.state_broadcast_secs)) {
+        sendCurrentStateMessage();
+        return DELAYED_INTERVAL;
+    }
+    return GPIO_POLLING_INTERVAL;
+}
+
+void DetectionSensorModule::sendDetectionMessage()
+{
+    LOG_DEBUG("Detected state change. Sending message\n");
+    char *message = new char[40];
+    sprintf(message, "%s detected", moduleConfig.detection_sensor.name);
+    meshtastic_MeshPacket *p = allocDataPacket();
+    p->want_ack = false;
+    p->decoded.payload.size = strlen(message);
+    memcpy(p->decoded.payload.bytes, message, p->decoded.payload.size);
+    if (moduleConfig.detection_sensor.send_bell && p->decoded.payload.size < meshtastic_Constants_DATA_PAYLOAD_LEN) {
+        p->decoded.payload.bytes[p->decoded.payload.size] = 7;        // Bell character
+        p->decoded.payload.bytes[p->decoded.payload.size + 1] = '\0'; // Bell character
+        p->decoded.payload.size++;
+    }
+    LOG_INFO("Sending message id=%d, dest=%x, msg=%.*s\n", p->id, p->to, p->decoded.payload.size, p->decoded.payload.bytes);
+    lastSentToMesh = millis();
+    service.sendToMesh(p);
+}
+
+void DetectionSensorModule::sendCurrentStateMessage()
+{
+    char *message = new char[40];
+    sprintf(message, "%s state: %i", moduleConfig.detection_sensor.name, hasStateChanged());
+
+    meshtastic_MeshPacket *p = allocDataPacket();
+    p->want_ack = false;
+    p->decoded.payload.size = strlen(message);
+    memcpy(p->decoded.payload.bytes, message, p->decoded.payload.size);
+    LOG_INFO("Sending message id=%d, dest=%x, msg=%.*s\n", p->id, p->to, p->decoded.payload.size, p->decoded.payload.bytes);
+    lastSentToMesh = millis();
+    service.sendToMesh(p);
+}
+
+bool DetectionSensorModule::hasStateChanged()
```

Comment:
Just a question if I understood correctly, because the term "has changed" seems confusing to me. To determine if sth. has changed you normally would need to compare it with the **old state**, but in this function the actual state is compared against the preset high or low. For motion this may make sense, but what about other types of sensors (e.g. water level). Should the message only be send if a "change" has been detected or periodically as long as the actual value differs from the preset?

Reply:
I agree. I think hasDetectionEvent() is a better description of what I'm looking for, because I don't actually want to be in the business of tracking previous states.